DOC-1936: Note SR authorization is on by default for Redpanda Cloud (#1694)
redpanda-data/documentation: docs/deploy/deployment-option/self-hosted/manual/production/production-deployment-automation.mdx

    | `distro` | Linux distribution to install. (This affects the `distro_` variables.) Default: `ubuntu-focal` |
    | `distro_ami` | AWS AMI to use for each available distribution. These must be changed according to the chosen AWS region. |
    | `distro_ssh_user` | Username used to SSH into the created EC2 instances. |

  </TabItem>

  <TabItem value="gcp" label="GCP">

  1. In the `deployment-automation` folder, change into the `gcp` directory:

    ```bash
    cd gcp
    ```

  2. You need an existing subnet in which to deploy the VMs. The subnet's attached firewall should allow inbound traffic on ports 22, 3000, 8082, 8888, 8889, 9090,  9092, 9644, and 33145. This module adds the `rp-node` tag to the deployed VMs, which can be used as the target tag for the firewall rule.

  3. Initialize Terraform:

    ```bash
    terraform init
    ```

  4. Create the cluster: 
 
    `terraform apply`

    The following example creates a three-broker cluster using the subnet named `redpanda-cluster-subnet`:

    ```bash
    terraform apply -var nodes=3 -var subnet=redpanda-cluster-subnet -var public_key_path=~/.ssh/id_rsa.pub -var ssh_user=$USER
    ```
   

>     Configuration options:

Reviewer: it might make sense to mention that this is a subset of available configuration options and link to the full list ([here](https://registry.terraform.io/modules/redpanda-data/redpanda-cluster/aws/latest?tab=inputs)) again.

     - If you don't have a default VPC defined, then you need to set a VPC ID and subnet ID.
     - If the public key path isn't the default `~/.ssh/id_rsa.pub`, then you need to set it. 
    
      ```bash
      terraform apply -var='public_key_path=~/.ssh/id_rsa.pub' -var='subnet_id=<subnet-id>' -var='vpc_id=<vpc-id>'
      ```
    
    Configuration options (for the full list, see the [Terraform module](https://registry.terraform.io/modules/redpanda-data/redpanda-cluster/aws/latest)):

    | Property      | Description |
    | ----------- | ----------- |
    | `aws_region` | The AWS region to use for deploying the infrastructure. Default: `us-west-2` |
    | `nodes` | The number of brokers to base the cluster on. Default: `3` |
    | `enable_monitoring` | Creates a Prometheus/Grafana instance for monitoring the cluster. Default: `true` |
    | `instance_type` | The instance type on which Redpanda is deployed. Default: `i3.8xlarge` |
    | `prometheus_instance_type` | The instance type on which Prometheus and Grafana are deployed. Default: `c5.2xlarge` |
    | `public_key_path` | Path to the public key of the keypair used to access the brokers. Default: `~/.ssh/id_rsa.pub` |

>     | `distro` | Linux distribution to install. (This affects the `distro_` variables.) Default: `ubuntu-focal` |

Reviewer 2: We did that before. Let's not go back to that. This is so much easier compared to surfing for the right region-specific ami name.

>     | `distro_ami` | AWS AMI to use for each available distribution. These must be changed according to the chosen AWS region. |

Reviewer 2: this should note that this is an override for whatever is set in distro. distro will automatically choose the latest ami version for a given OS in a given region, but if the user needs a specific local one, they can set distro_ami accordingly. 

also, if they set distro_ami, they must also set distro_ssh_user.

    ```bash
    cd gcp
    ```

  2. You need an existing subnet in which to deploy the VMs. The subnet's attached firewall should allow inbound traffic on ports 22, 3000, 8082, 8888, 8889, 9090,  9092, 9644, and 33145. This module adds the `rp-node` tag to the deployed VMs, which can be used as the target tag for the firewall rule.

  3. Initialize Terraform:

    ```bash
    terraform init
    ```

  4. Create the cluster: 
 
    `terraform apply`

    The following example creates a three-broker cluster using the subnet named `redpanda-cluster-subnet`:

    ```bash
    terraform apply -var nodes=3 -var subnet=redpanda-cluster-subnet -var public_key_path=~/.ssh/id_rsa.pub -var ssh_user=$USER
    ```
   
    Configuration options (for the full list, see the [Terraform module](https://registry.terraform.io/modules/redpanda-data/redpanda-cluster/aws/latest)):
   
    | Property      | Description |
    | ----------- | ----------- |
    | `region` | The region to use for deploying the infrastructure. Default: `us-west1` |
    | `zone` | The region's zone to deploy the infrastructure. Default: `a` |
    | `subnet` | The name of an existing subnet to deploy the infrastructure. |

>     | `nodes` | The number of brokers to base the cluster on. Keep in mind that one broker is used as a monitoring broker. Default: `1`|

Reviewer 2: @gene-redpanda I thought we separated out the monitoring node to it's own instance, separate from the broker instances?

Reviewer: Correct, that sentence has been wrong since before I got here 😂

Author: I'll remove "Keep in mind that one node is used as a monitoring node."

    | `disks` | The number of local disks to deploy on each machine. Default = `1`|
    | `image` | The OS image running on the VMs. Default: `ubuntu-os-cloud/ubuntu-1804-lts` |
    | `machine_type` | The machine type. Default: `n2-standard-2`|
    | `public_key_path` | Path to the public key of the keypair used to access the brokers. |
    | `ssh_user` | The SSH user. Must match the one in the public SSH key's comments. |

  </TabItem>

</Tabs>

> 5. Now you can install Redpanda either with Ansible or with the [Redpanda installation binary](../../production/production-deployment).

Reviewer 2: If this doc is focused on automated and repeatable installs, we shouldn't mention the Redpanda installation binary.

## Use Ansible to install Redpanda

> 1. Set required Ansible variables:

Reviewer 2: need to make sure the user has gone back into the top-level directory for deployment-automation. I think if they're following along in the doc, they'd still be somewhere in a terraform directory when they got to this phase

  ```bash

>   export CLOUD_PROVIDER=<cloud-provider>

Reviewer 2: We should note what the available cloud provider values are for this so it's clear to the end user.

  export ANSIBLE_COLLECTIONS_PATHS=${PWD}/artifacts/collections
  export ANSIBLE_ROLES_PATH=${PWD}/artifacts/roles
  export ANSIBLE_INVENTORY=${PWD}/${CLOUD_PROVIDER}/hosts.ini
  ```
  
1. Install the roles required by Ansible:

  ```bash
  ansible-galaxy install -r ansible/requirements.yml
  ```

### Configure a hosts file

If you used Terraform to deploy the instances, the `hosts.ini` is configured automatically in the [`artifacts`](https://github.com/redpanda-data/deployment-automation/tree/main/artifacts) directory. 

If you didn't use Terraform, then you must manually update the `[redpanda]` section. When you open the file, you see something like the following:

```ini
[redpanda]
ip ansible_user=ssh_user ansible_become=True private_ip=pip id=0
ip ansible_user=ssh_user ansible_become=True private_ip=pip id=1

[monitor]
ip ansible_user=ssh_user ansible_become=True private_ip=pip id=1
```

Under the `[redpanda]` section, replace the following:

| Property      | Description |
| ----------- | ----------- |
| `ip` | The public IP address of the machine. |
| `ansible_user` | The username for Ansible to use to SSH to the machine. |

> | `private_ip` | The private IP address of the machine. This could be the same as the public IP address. |

Reviewer 2: @gene-redpanda  private_ip is optional if there's only one ip address for the host right?

Reviewer: I'd keep it

> | `id` | The node ID of the Redpanda instance. This must be unique for each host. |

Reviewer 2: is id still needed? I thought this requirement was removed since it was originally used for defining the internal broker id

Reviewer: Can be removed

> You can add additional properties to configure features like rack awareness and Tiered Storage. The `[monitor]` section is only relevant if you have Prometheus and Grafana installed on a given host. If you don't want to have this deployed, then remove the `[monitor]` section.

Reviewer 2: I'd reword this slightly

> The `[monitor]` section is only required if you want the Redpanda Ansible playbooks to install and configure a basic Prometheus and Grafana setup for observability. If you have a centralized monitoring setup or don't require monitoring, then remove this section.

### Run a playbook

> Use the [Redpanda Ansible Collection](https://galaxy.ansible.com/redpanda/cluster) to build a Redpanda cluster. The recommended Redpanda playbook enables TLS encryption and Tiered Storage. 

Reviewer 3: on the destination page, the heading reads "Ansible Collection for Redpanda"

Author: thanks, fixed!

> Use the [Redpanda Ansible Collection](https://galaxy.ansible.com/redpanda/cluster) to build a Redpanda cluster. The recommended Redpanda playbook enables TLS encryption and Tiered Storage. 

Reviewer 2: @gene-redpanda Thinking a bit more about deployments: TS will only be configurable if the user has access to an object store. When deploying via Ansible-only, we probably can't assume that because they likely will be on-prem with no access to an s3 compatible bucket.

> If you prefer, you can download the modules and required roles and create your own playbook. For example, if you want to handle your own data directory, you can toggle that part off, and Redpanda just makes sure the permissions are correct. If you want to generate your own security certificates, you can. 

Reviewer 3: ```suggestion
If you prefer, you can download the modules and required roles and create your own playbook. For example, if you want to handle your own data directory, you can toggle that part off, and Redpanda ensures that the permissions are correct. If you want to generate your own security certificates, you can. 
```

> To run the recommended Redpanda playbook and have a fully-running Redpanda cluster in one command, run:

Reviewer 2: > To install and start a Redpanda cluster in one command with the Redpanda playbook, run:

Can we restate to that?

 ```bash

>  ansible-playbook --private-key <your-private-key> -i hosts.ini -v ansible/playbooks/provision-basic-cluster.yml

Reviewer 2: I think `-i hosts.ini` is not needed if we previously set the inventory location in a shell variable

we should also add a note that the private key used here is the one that corresponds to the public key loaded into the distro_user's ssh configuration.

 ```

To use your own playbook, replace `provision-basic-cluster.yml` with your playbook name. 

:::note

> When you use a playbook to create a cluster, you should also use the playbook for subsequent operations, like upgrades.

Reviewer 2: Good place to call out that the Ansible modules correctly (and safely) handle doing rolling upgrades, but you must still comply with the version path requirements for RP (there's a doc link we can add in here talking about that)

:::

#### Custom configuration

> You can specify any available Redpanda configuration value, or set of values, by passing a JSON dictionary as an Ansible `extra-var`. These values are spliced with the calculated configuration and only override the values that you specify. Values must be unset manually with `rpk`. There are two sub-dictionaries you can specify: `redpanda.cluster` and `redpanda.node`. For more information, see [Cluster Configuration Properties](../../../../../../reference/cluster-properties) and [Node Configuration Properties](../../../../../../reference/node-properties). 

Reviewer 2: @gene-redpanda do we have an example that doesn't pass this in as a json data from the command line?

Reviewer: You can take the value you'd put for the command line and put it into a hosts file for exactly the same result.

```bash
export JSONDATA='{"cluster":{"auto_create_topics_enabled":"true"},"node":{"developer_mode":"false"}}'
ansible-playbook ansible/<playbook-name>.yml --private-key artifacts/testkey -e redpanda="${JSONDATA}"
```

:::note
Adding whitespace to the JSON breaks configuration merging. 
:::

Use `rpk` and standard Kafka tools to produce and consume from the Redpanda cluster.

#### Configure Grafana

> To deploy a Grafana node, you must have a `[monitor]` section in your hosts file. The Grafana URL is `http://<grafana-host>:3000/login`.

Reviewer 2: This would also deploy prometheus in combination.

This would be a good section to include a statement about Redpanda emitting Prometheus metrics that can be scrapped with a central collector. There's no need to use our playbooks to deploy prom/graf if they already have something setup.

To run the `deploy-prometheus-grafana.yml` playbook:

```bash
ansible-playbook ansible/deploy-prometheus-grafana.yml \

> -i hosts.ini \

Reviewer 2: not needed if we already previously defined the inventory file in the shell variable

--private-key '<path-to-a-private-key-with-ssh-access-to-the-hosts>'
```

#### Build the cluster with TLS enabled

Configure TLS with externally-provided and signed certificates. Then run the `provision-tls-cluster` playbook, specifying the certificate locations on new hosts. You can either pass the variables in the command line or edit the file and pass them there. Consider whether you want public access to the Kafka API and Admin API endpoints. For example:

```bash
ansible-playbook ansible/provision-tls-cluster.yml \

> -i hosts.ini \

Reviewer 2: not needed if we already previously defined the inventory file in the shell variable

--private-key '<path-to-a-private-key-with-ssh-access-to-the-hosts>' \

> --extra-vars create_demo_certs=false \

Reviewer 2: @gene-redpanda we should probably flip this so it's always false so we don't have to specify it every time. creating demo certs should be a special case.

--extra-vars advertise_public_ips=false \ 
--extra-vars handle_certs=false \
--extra-vars redpanda_truststore_file='<path-to-ca.crt-file>' 
```

It is important to use a signed certificate from a valid CA for production environments. The playbook uses locally-signed certificates that are not recommended for production use. Provide a valid certificate using these variables:

```bash
redpanda_certs_dir: /etc/redpanda/certs
redpanda_csr_file: "{{ redpanda_certs_dir }}/node.csr"
redpanda_key_file: "{{ redpanda_certs_dir }}/node.key"
redpanda_cert_file: "{{ redpanda_certs_dir }}/node.crt"
redpanda_truststore_file: "{{ redpanda_certs_dir }}/truststore.pem"
```

For testing, you could deploy a local CA to generate private keys and signed certificates:

```bash
ansible-playbook ansible/provision-tiered-storage-cluster.yml \

> -i hosts.ini \

Reviewer 2: not needed if we already previously defined the inventory file in the shell variable

--private-key '<path-to-a-private-key-with-ssh-access-to-the-hosts>'
```

#### Add brokers to an existing cluster

To add brokers to a cluster, you must add them to the hosts file and run the relevant playbook again. You can add `skip_node=true` to the existing hosts to avoid the playbooks being rerun on them.

#### Upgrade a cluster

The playbook is designed to be idempotent, so it should be suitable for running as part of a CI/CD pipeline or through Ansible Tower. The playbook upgrades the packages and then performs a [rolling upgrade](../../../../../../manage/cluster-maintenance/rolling-upgrade/#perform-a-rolling-upgrade) across the cluster.

:::note notes
- Because edits made to properties outside of the playbook may be overwritten by this procedure, incorporate variables into your `hosts.ini` file.
- Test that your upgrade path is safe before using it in production.

> :::

Reviewer 2: this is a good general note that we should have at the beginning of the ansible section.

the challenge with using multiple --extra-vars command flags is that you have to remember to use them every time., which is hard for repeatability. Tower makes this easier, I think.

but knowing that they can stick these in the hosts.ini for every host (or even better, if we could pull them in from a single include_vars file)

@gene-redpanda  thoughts?

Reviewer: Yeah I've mentioned this a couple times as an important detail so am in agreement

> To upgrade a cluster, run the playbook with a specific target version. If the target version is higher than the currently-installed version, then the cluster is upgraded and restarted automatically:

Reviewer 2: Can we rephrase slightly

> is upgraded and safely restarted using a rolling restart that operates on one broker at a time.

This is also a good place to stick a link to our general version upgrade docs talking about what versions you can upgrade between (e.g. 22.3->23.1, but not 22.2->23.1 since that skips a major)

```bash
ansible-playbook --private-key ~/.ssh/id_rsa ansible/<playbook-name>.yml -i hosts.ini -e redpanda_version=22.3.10-1 
```

By default, the playbook selects the latest version of the Redpanda packages, but an upgrade is only performed if the `redpanda_install_status` variable is set to `latest`:

```bash

> ansible-playbook --private-key ~/.ssh/id_rsa ansible/<playbook-name>.yml -i hosts.ini -e redpanda_install_status=latest 

Reviewer 2: does redpanda_install_status have to be `latest` or just any value? I think this corresponds back to the Ansible `state:` value right?

```

You can upgrade clusters with SASL authentication by specifying the `redpanda_rpk_opts` variable to include the username and password of a superuser or appropriately-privileged user. For example:

```bash
ansible-playbook --private-key ~/.ssh/id_rsa ansible/<playbook-name>.yml -i hosts.ini --extra-vars=redpanda_install_status=latest --extra-vars "{

> \"redpanda_rpk_opts\": \"--user ${MY_USER} --password ${MY_USER_PASSWORD}\"

Reviewer 2: @gene-redpanda is there a better way to define the extra vars here that doesn't require backslashes?

Reviewer: This example is incorrect. Here is a correct example:

```shell
export JSONDATA='{"cluster":{"auto_create_topics_enabled":"true"},"node":{"developer_mode":"false"}}'
ansible-playbook ansible/<<<PLAYBOOK NAME>>>.yml --private-key artifacts/testkey -e redpanda="${JSONDATA}"
```

Author: @gene-redpanda please confirm that this looks correct now!

Reviewer: Yep looks great!

| `redpanda_rpk_opts` |  | Command line options to be passed to instances where `rpk` is used on the playbook; for example, superuser credentials can be specified as `--user myuser --password mypassword`. |
| `redpanda_install_status` | `present` | If `redpanda_version` is set to latest, changing `redpanda_install_status` to latest causes an upgrade; otherwise, the currently-installed version remains. |
| `redpanda_data_directory` | `/var/lib/redpanda/data` | Path where Redpanda keeps its data. |
| `redpanda_key_file` | `/etc/redpanda/certs/node.key` | TLS: Path to private key. |
| `redpanda_cert_file` | `/etc/redpanda/certs/node.crt` | TLS: Path to signed certificate. |
| `redpanda_truststore_file` | `/etc/redpanda/certs/truststore.pem` | TLS: Path to truststore. |
| `tls` | `false` | Set to true to configure Redpanda to use TLS. This can be set on each broker, although this may lead to errors configuring `rpk`. |
| `skip_node` |	`false`	| Node configuration to prevent the redpanda_broker role being applied to this specific broker. Use carefully when adding new brokers to avoid existing brokers from being reconfigured. |

> | `restart_node` | `false` | Node configuration to prevent Redpanda brokers from being restarted after updating. Use with care: this can cause `rpk` to be reconfigured but the broker not be restarted and therefore be in an inconsistent state. |

Reviewer 3: ```suggestion
| `restart_node` | `false` | Node configuration to prevent Redpanda brokers from being restarted after updating. Use with care: this can cause `rpk` to be reconfigured but the broker is not restarted and therefore is in an inconsistent state. |
```

| `rack` | `undefined` | Node configuration to enable rack awareness. Rack awareness is enabled cluster-wide if at least one broker has the rack variable set. |
| `tiered_storage_bucket_name` |  | Set bucket name to enable Tiered Storage. |

> | `aws_region` |  | The region to be used if Tiered Storage is enabled. |

Author: @hcoyote The doc will be the single source or truth. After this merges, we can remove the duplicate content from the repo readme.

> | `aws_region` |  | The region to be used if Tiered Storage is enabled. |

Reviewer 2: Works for me!

### Troubleshooting

On Mac OS X, Python may be [unable to fork workers](https://stackoverflow.com/questions/50168647/multiprocessing-causes-python-to-crash-and-gives-an-error-may-have-been-in-progr). If you see something like this:

```bash
ok: [34.209.26.177] => {“changed”: false, “stat”: {“exists”: false}}
objc[57889]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called.
objc[57889]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
ERROR! A worker was found in a dead state
```

> To resolve this, try setting an environment variable: 

Reviewer 3: ```suggestion
Try setting an environment variable to resolve the error: 
```

## Prerequisites

1. Install Terraform following the [Terraform documentation](https://learn.hashicorp.com/tutorials/terraform/install-cli).

2. Install Ansible following the [Ansible documentation](https://docs.ansible.com/ansible/latest/installation_guide/intro_installation.html). Different operating systems may have specific Ansible dependencies.

3. Clone the [`deployment-automation` GitHub repository](https://github.com/redpanda-data/deployment-automation/):

 ```bash
 git clone https://github.com/redpanda-data/deployment-automation.git
 ```

4. Change into the directory:

 ```bash
 cd deployment-automation
 ```

## Use Terraform to set up infrastructure

<Tabs>
  <TabItem value="aws" label="AWS" default>

  The recommended [Terraform module for Redpanda](https://registry.terraform.io/modules/redpanda-data/redpanda-cluster/aws/latest) deploys virtual machines on AWS EC2. To create an AWS Redpanda cluster, review the [default variables](https://github.com/redpanda-data/deployment-automation/blob/main/aws/main.tf) and make any edits necessary for your environment. 

  1. In the `deployment-automation` folder, change into the `aws` directory:

    ```bash

>     cd aws/standard

Reviewer 4: Is this supposed to be `aws/standard` or just `aws`? I believe I've always used aws/standard but would like to confirm with @gene-redpanda.

Reviewer: aws/standard is what it will be in the new version so I'd like to be correct with that @deniscoady

Reviewer 4: > aws/standard is what it will be in the **new version** so I'd like to be correct with that @deniscoady

When is this new version going to be publicly available? I just checked the repo and it's still just `aws/`. I want to make sure that the docs reflect what customers can do now.

Author: agreed, docs are public as soon as we merge, and because of our docs migration we want to merge this by Friday July 28. @gene-redpanda is it OK if I change this back to `aws/`, and we'll update to `aws/standard` when it changes?

Reviewer: That's fine then. I will say it will change probably in the next week -- we're working thru the review process rn.

Author: Sorry to pile on, but this existing section may need updates too: https://docs.redpanda.com/docs/deploy/deployment-option/self-hosted/manual/high-availability/#single-az-example

Reviewer 4: If it'll be reviewed and ready by next week then I think it's okay. 

cc: @twmb

Author: Confirmed that we'll merge this now to get this update included. We'll open a new issue to update `aws/standard` when it's ready, as well as any updates required here: https://docs.redpanda.com/docs/deploy/deployment-option/self-hosted/manual/high-availability/#single-az-example.

| `ansible_user` | The username for Ansible to use to SSH to the machine. |
| `private_ip` | The private IP address of the machine. This could be the same as the public IP address. |

You can add additional properties to configure features like rack awareness and Tiered Storage. 

The `[monitor]` section is only required if you want the playbook to install and configure a basic Prometheus and Grafana setup for observability. If you have a centralized monitoring setup or if you don't require monitoring, then remove this section.

### Run a playbook

Use the [Ansible Collection for Redpanda](https://galaxy.ansible.com/redpanda/cluster) to build a Redpanda cluster. The recommended Redpanda playbook enables TLS encryption and Tiered Storage. 

If you prefer, you can download the modules and required roles and create your own playbook. For example, if you want to handle your own data directory, you can toggle that part off, and Redpanda ensures that the permissions are correct. If you want to generate your own security certificates, you can. 

To install and start a Redpanda cluster in one command with the Redpanda playbook, run:

 ```bash

>  ansible-playbook --private-key <your-private-key> -i hosts.ini -v ansible/playbooks/provision-basic-cluster.yml

Reviewer: ```suggestion
 ansible-playbook --private-key <your-private-key> -v ansible/playbooks/provision-basic-cluster.yml
```

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> If you use automation tools like Terraform and Ansible in your environment, you can use them to easily provision a Redpanda cluster for production. Terraform can set up the infrastructure and output a properly-formatted `hosts.ini` file, and Ansible can use that `hosts.ini` file as input to install Redpanda. Or, you can supply your own hosts file but still use Ansible for installation.

Reviewer: It might make sense to show what our hosts file looks like here. Example: https://github.com/redpanda-data/deployment-automation#populating-a-hosts-file

Author: Please confirm if we need anything more in our [hosts example](https://deploy-preview-1694--redpanda-documentation.netlify.app/docs/deploy/deployment-option/self-hosted/manual/production/production-deployment-automation/#configure-a-hosts-file).

  2. You need an existing subnet to deploy the VMs into. The subnet's attached firewall should allow inbound traffic on ports 22, 3000, 8082, 8888, 8889, 9090,  9092, 9644, and 33145. This module adds the `rp-node` tag to the deployed VMs, which can be used as the target tag for the firewall rule.

  3. Initialize Terraform:

    ```bash
    terraform init
    ```

  4. Create the cluster: 
 

>     `terraform apply`

Reviewer: This is a duplicate of 52-65 now

    The following example creates a three-broker cluster using the subnet named `redpanda-cluster-subnet`:

    ```bash
    terraform apply -var nodes=3 -var subnet=redpanda-cluster-subnet -var public_key_path=~/.ssh/id_rsa.pub -var ssh_user=$USER

>     ```

Reviewer: This is now a duplicate of 52-65

  2. You need an existing subnet to deploy the VMs into. The subnet's attached firewall should allow inbound traffic on ports 22, 3000, 8082, 8888, 8889, 9090,  9092, 9644, and 33145. This module adds the `rp-node` tag to the deployed VMs, which can be used as the target tag for the firewall rule.

  3. Initialize Terraform:

    ```bash
    terraform init
    ```

  4. Create the cluster: 
 

>     `terraform apply`

Reviewer: you will need a vpc id and subnet id for this to work if you don't have a default vpc defined. the public key will also need to be set if it isn't the default ~/.ssh/id_rsa.pub

```shell
terraform apply -var='public_key_path=~/.ssh/id_rsa.pub' -var='subnet_id=<<<SUBNET_ID>>>' -var='vpc_id=<<<VPC_ID>>>'
```

| `ansible_user` | The username for Ansible to use to SSH to the machine. |
| `private_ip` | The private IP address of the machine. This could be the same as the public IP address. |
| `id` | The node ID of the Redpanda instance. This must be unique for each host. |

You can add additional properties to configure features like rack awareness and Tiered Storage. The `[monitor]` section is only relevant if you have Prometheus and Grafana installed on a given host. If you don't want to have this deployed, then remove the `[monitor]` section.

### Run a playbook

Use the [Redpanda Ansible Collection](https://galaxy.ansible.com/redpanda/cluster) to build a Redpanda cluster. The recommended Redpanda playbook enables TLS encryption and Tiered Storage. 

If you prefer, you can download the modules and required roles and create your own playbook. For example, if you want to handle your own data directory, you can toggle that part off, and Redpanda just makes sure the permissions are correct. If you want to generate your own security certificates, you can. 

To run the recommended Redpanda playbook and have a fully-running Redpanda cluster in one command, run:

 ```bash

>  ansible-playbook --private-key <your-private-key> -i hosts.ini -v ansible/playbooks/provision-node.yml

Reviewer: we no longer use provision-node.yml as it confused people. the most basic playbook is now provision-basic-cluster.yml

 ```

> To use your own playbook, replace `provision-node.yml` with your playbook name. 

Reviewer: same comment

:::note

> If you use a playbook to create a cluster, also use it for any subsequent operations, like upgrades.

Reviewer: ```suggestion
If you use a playbook to create a cluster also use it for any subsequent operations like upgrades.
```

To run the `deploy-prometheus-grafana.yml` playbook:

```bash
ansible-playbook ansible/deploy-prometheus-grafana.yml \
-i hosts.ini \
--private-key '<path-to-a-private-key-with-ssh-access-to-the-hosts>'
```

#### Build the cluster with TLS enabled

Configure TLS with externally-provided and signed certificates. Then run the `provision-tls-cluster` playbook, specifying the certificate locations on new hosts. You can either pass the variables in the command line or edit the file and pass them there. Consider whether you want public access to the Kafka API and Admin API endpoints. For example:

```bash
ansible-playbook ansible/provision-tls-cluster.yml \
-i hosts.ini \
--private-key '<path-to-a-private-key-with-ssh-access-to-the-hosts>' \
--extra-vars create_demo_certs=false \
--extra-vars advertise_public_ips=false \ 
--extra-vars handle_certs=false \
--extra-vars redpanda_truststore_file='<path-to-ca.crt-file>' 
```

For testing, you could deploy a local CA to generate private keys and signed certificates:

```bash
ansible-playbook ansible/provision-tiered-storage-cluster.yml \
-i hosts.ini \
--private-key '<path-to-a-private-key-with-ssh-access-to-the-hosts>'
```

:::important

> Use a signed certificate from a valid CA for production environments.

Reviewer: ```suggestion
Use a signed certificate from a valid CA for production environments. Our playbook defaults to using locally signed certs that are not recommended for production use and are intended for demo purposes only. You can provide a valid cert using these variables:

redpanda_certs_dir: /etc/redpanda/certs
redpanda_csr_file: "{{ redpanda_certs_dir }}/node.csr"
redpanda_key_file: "{{ redpanda_certs_dir }}/node.key"
redpanda_cert_file: "{{ redpanda_certs_dir }}/node.crt"
redpanda_truststore_file: "{{ redpanda_certs_dir }}/truststore.pem"

```

:::

#### Add brokers to an existing cluster

To add brokers to a cluster, you must add them to the hosts file and run the relevant playbook again. You can add `skip_node=true` to the existing hosts to avoid the playbooks being rerun on them.

#### Upgrade a cluster

The playbook is designed to be idempotent, so it should be suitable for running as part of a CI/CD pipeline or through Ansible Tower. The playbook upgrades the packages and then performs a [rolling upgrade](../../../../../../manage/cluster-maintenance/rolling-upgrade/#perform-a-rolling-upgrade) across the cluster.

:::note notes
- Because edits made to properties outside of the playbook may be overwritten by this procedure, these settings should be incorporated as `--extra-vars`. For example, `--extra-vars=enable_tls=true`. See [Custom configuration](#custom-configuration).

> - Test that your upgrade path is safe before using it in production.

Reviewer: It is generally a good idea with ansible to incorporate variables into the hosts.ini file rather than including them as extra vars so that future runs will be executed the same way as the first one.

Author: @gene-redpanda: Can you pls confirm that we should replace the first bullet here about incorporate settings as `--extra-vars` with: 

Because edits made to properties outside of the playbook may be overwritten by this procedure, incorporate variables into the `hosts.ini` file.

Reviewer: Yeah that's fine

    | `zone` | The region's zone to deploy the infrastructure. Default: `a` |
    | `subnet` | The name of an existing subnet to deploy the infrastructure. |
    | `nodes` | The number of brokers to base the cluster on. Keep in mind that one broker is used as a monitoring broker. Default: `1`|
    | `disks` | The number of local disks to deploy on each machine. Default = `1`|
    | `image` | The OS image running on the VMs. Default: `ubuntu-os-cloud/ubuntu-1804-lts` |
    | `machine_type` | The machine type. Default: `n2-standard-2`|
    | `public_key_path` | Path to the public key of the keypair used to access the brokers. |
    | `ssh_user` | The ssh user. Must match the one in the public ssh key's comments. |

  </TabItem>

</Tabs>

5. Now you can install Redpanda either with Ansible or with the [Redpanda installation binary](../../production/production-deployment).

:::note
Non-AWS Terraform code is also included for [Azure](https://github.com/redpanda-data/deployment-automation/blob/main/azure/README.md) and [IBM Cloud](https://github.com/redpanda-data/deployment-automation/blob/main/ibm/README.md):
```bash
export AWS_ACCESS_KEY_ID=<your-key-ID>

> export AWS_SECRET_ACCESS_KEY=<your-secret-access-key>

Reviewer: GCP as well

Reviewer: It'd make sense to just call it "Terraform code" or "Terraform scripts" rather than "Non-AWS Terraform Code"

# Go to the cloud provider directory (aws, gcp, azure, or IBM)
cd azure
terraform init
terraform apply --auto-approve
```
:::

## Use Ansible to install Redpanda

If you haven't yet installed Ansible, do so following the [Ansible documentation](https://docs.ansible.com/ansible/latest/installation_guide/intro_installation.html). Depending on your system, you may also need to install Python packages, like SELinux or JMESPath.

1. Set required Ansible variables:

  ```bash
  export CLOUD_PROVIDER=<cloud-provider>
  export ANSIBLE_COLLECTIONS_PATHS=${PWD}/artifacts/collections
  export ANSIBLE_ROLES_PATH=${PWD}/artifacts/roles
  export ANSIBLE_INVENTORY=${PWD}/${CLOUD_PROVIDER}/hosts.ini
  ```

> 1. Clone the [`deployment-automation` GitHub repository](https://github.com/redpanda-data/deployment-automation/):

Reviewer: This was already done at line 20

   ```bash
   git clone https://github.com/redpanda-data/deployment-automation.git
   ```

1. Change into the directory:

  ```bash
  cd deployment-automation

>   ```

Reviewer: Same as above, already accomplished ~line 20

### Next steps

Now you can install Redpanda either with the [Redpanda installation binary](../../production/production-deployment) or with Ansible.

## Install Redpanda with Ansible
Now you can install Redpanda with Ansible or with the [Redpanda installation binary](../../production/production-deployment).

Whether you are using bare metal servers or Terraform, you can use Ansible to install Redpanda.
## Use Ansible to install Redpanda

### Prerequisites
Whether you're using bare metal servers or Terraform, you can use Ansible to install Redpanda. If you haven't yet installed Ansible, do so following the [Ansible documentation](https://docs.ansible.com/ansible/latest/installation_guide/intro_installation.html).

Install Ansible. See the [Ansible documentation](https://docs.ansible.com/ansible/latest/installation_guide/intro_installation.html).
1. Clone the `deployment-automation` GitHub repository:

> ### Clone the GitHub repository

Reviewer: It might be sensible to include some of the information from the Basic Usage section here: 

https://github.com/redpanda-data/deployment-automation#basic-usage

Author: I'm thinking this should stay in the readme, with a link to the updated doc page for more info. After the updated doc is published, I think almost everything on the readme can be removed except Basic Usage and Configuration Guide, but please confirm!

Reviewer: Well, the environment variables are essential for ansible to work. otherwise i agree

| Property      | Default value | Description |
| ----------- | ----------- | ----------- |
| `redpanda_organization` | `redpanda-test` | Set this to identify your organization in the asset management system. |
| `redpanda_cluster_id` | `redpanda` | Helps identify the cluster. |
| `advertise_public_ips` | `false` | Configure Redpanda to advertise the node's public IPs for client communication instead of private IPs. This enables using the cluster from outside its subnet. Note: This is not recommended for production deployments, because it means that your nodes will be public. 
| `grafana_admin_pass` | <your_secure_password> | Configure Grafana's admin user's password. |
| `ephemeral_disk` | `false` | Enable filesystem check for attached disk. This is useful when using attached disks in instances with ephemeral OS disks like Azure L Series. This allows a filesystem repair at boot time and ensures that the drive is remounted automatically after a reboot. |
| `redpanda_mode` | `production` | Enables hardware optimization. |
| `redpanda_admin_api_port` | `9644` |  |
| `redpanda_kafka_port` | `9092` |  |
| `redpanda_rpc_port` | `33145` |  |

> | `redpanda_use_staging_repo` | `false` | Enables access to unstable builds. |

Reviewer: this is now "is_using_unstable"

Author: is it `redpanda_is_using_unstable `or just `is_using_unstable`?

Reviewer: @micheleRP ```is_using_unstable: false```

| `restart_node` | `false` | Per-node config to prevent Redpanda brokers from being restarted after updating. Use with care: this can cause `rpk` to be reconfigured but the node not be restarted and therefore be in an inconsistent state. |
| `rack` | `undefined` | Per-node config to enable rack awareness. Rack awareness is enabled cluster-wide if at least one node has the rack variable set. |
| `tiered_storage_bucket_name` |  | Set bucket name to enable Tiered Storage. |
| `aws_region` |  | The region to be used if Tiered Storage is enabled. |

#### Custom configuration

You can specify any available Redpanda configuration value (or set of values) by passing a JSON dictionary as an Ansible `extra-var`. These values are spliced with the calculated configuration and only override those values that you specify. There are two sub-dictionaries that you can specify: `redpanda.cluster` and `redpanda.node`. For more information, see [Cluster Configuration Properties](../../../../../../reference/cluster-properties) and [Node Configuration Properties](../../../../../../reference/node-properties).

> 

Reviewer: we should probably also note that removing a value from here does not null it in redpanda. There's a technical explanation for why, but it boils down to ansible not having state so not knowing that a value has been removed. Since we can't act on the lack of a thing (and haven't implemented a way for users to provide a value to be removed from the config yet) any values that should be unset must be unset manually by using RPK

```bash
export JSONDATA='{"cluster":{"auto_create_topics_enabled":"true"},"node":{"developer_mode":"false"}}'
ansible-playbook ansible/<playbook-name>.yml --private-key artifacts/testkey -e redpanda="${JSONDATA}"
```

:::note
Adding whitespace breaks configuration merging. 
:::

Use `rpk` and standard Kafka tools to produce/consume from the Redpanda cluster and access the Grafana installation on the monitor host. The Grafana URL is `http://<grafana host>:3000/login`.

#### Configure Grafana

To deploy a Grafana node, you must have a `[monitor]` section in your hosts file. Then run the `deploy-prometheus-grafana.yml` playbook:

```bash
ansible-playbook ansible/deploy-prometheus-grafana.yml \
-i hosts.ini \
--private-key '<path to a private key with ssh access to the hosts>'
```

#### Build the cluster with TLS enabled

> Configure TLS with externally-provided and signed certificates (possibly through a corporate-provided Certmonger). Then run the `provision-tls-cluster` playbook, specifying the certificate locations on new hosts. You can either pass the variables in the command line or edit the file and pass them there. Consider whether you want public access to the Kafka API and Admin API endpoints. For example:

Reviewer: cert manager not Certmonger

Author: @gene-redpanda I'm not sure about the right terminology for this. Which of these is correct? 

Configure TLS with externally-provided and signed certificates (possibly through a corporate-provided cert-manager).
or
Configure TLS signed certificates (possibly through cert-manager).
or 
Configure TLS signed certificates (possibly generated by a corporate CA).
or 
Configure TLS signed certificates (through cert-manager or a corporate CA).

Reviewer: we can safely delete this parenthetical:  "(possibly through a corporate-provided Certmonger)"

```bash
ansible-playbook ansible/provision-tls-cluster.yml \
-i hosts.ini \
--private-key '<path to a private key with ssh access to the hosts>' \
--extra-vars create_demo_certs=false \
--extra-vars advertise_public_ips=false \ 
--extra-vars handle_certs=false \
--extra-vars redpanda_truststore_file='<path to ca.crt file>' 
```

#### Add nodes to an existing cluster

To add nodes to a cluster, you must add them to the hosts file and run the relevant playbook again. You can add `skip_node=true` to the existing hosts to avoid the playbooks being rerun on existing nodes.

#### Upgrade a cluster

The playbook is designed to be idempotent, so it should be suitable for running as part of a CI/CD pipeline or through Ansible Tower. Upgrade support is built in, and the playbook is capable of upgrading the packages and then performing a rolling upgrade across the cluster.

:::note

> Because any changes made to properties outside of the playbook may be overwritten by this procedure, these settings should be incorporated as part of the provided `--extra-vars`. For example, `--extra-vars=enable_tls=true`.

Reviewer: might want to also reference the information on line 252 again here.

:::

To enact an upgrade on a cluster, run the playbook with a specific target version. If the target version is higher than the currently installed version, then the cluster is upgraded and restarted automatically:

> 

Reviewer: "To upgrade a cluster" instead of "to enact an upgrade on a cluster"

Reviewer: Might want to add a warning about testing this in staging first

```bash
ansible-playbook --private-key ~/.ssh/id_rsa ansible/<playbook-name>.yml -i hosts.ini -e redpanda_version=22.3.10-1 
```

> By default, the playbook selects the latest version of the Redpanda packages, but an upgrade is only enacted if the `redpanda_install_status` variable is set to `latest`:

Reviewer: ```suggestion
By default the playbook selects the latest version of the Redpanda packages, but an upgrade is only enacted if the `redpanda_install_status` variable is set to `latest`:
```

```bash
ansible-playbook --private-key ~/.ssh/id_rsa ansible/<playbook-name>.yml -i hosts.ini -e redpanda_install_status=latest 
```

It's possible to upgrade clusters where SASL authentication has been turned on. For this, you need to additionally specify the `redpanda_rpk_opts` variable to include to username and password or a superuser or appropriately privileged user. For example:

```bash
ansible-playbook --private-key ~/.ssh/id_rsa ansible/<playbook-name>.yml -i hosts.ini --extra-vars=redpanda_install_status=latest --extra-vars "{
\"redpanda_rpk_opts\": \"--user ${MY_USER} --password ${MY_USER_PASSWORD}\"
}"
```

Similarly, you can put the `redpanda_rpk_opts` into a YAML file protected with Ansible vault.

```bash
ansible-playbook --private-key ~/.ssh/id_rsa ansible/<playbook-name>.yml -i hosts.ini --extra-vars=redpanda_install_status=latest --extra-vars @vault-file.yml --ask-vault-pass
```

##### Migrate to the AWS module

> Redpanda previously delivered AWS code as simple in-place Terraform code, but it's now converted to a module. To migrate from the old AWS code:

Reviewer: ```suggestion
Redpanda previously delivered AWS code as simple in-place Terraform code, but it is now converted to a module. To migrate from the old AWS code:
```

Nitpicking like crazy 😅

| `rack` | `undefined` | Per-node config to enable rack awareness. Rack awareness is enabled cluster-wide if at least one node has the rack variable set. |
| `tiered_storage_bucket_name` |  | Set bucket name to enable Tiered Storage. |
| `aws_region` |  | The region to be used if Tiered Storage is enabled. |

#### Custom configuration

You can specify any available Redpanda configuration value (or set of values) by passing a JSON dictionary as an Ansible `extra-var`. These values are spliced with the calculated configuration and only override the values that you specify. Values must be unset manually with `rpk`. There are two sub-dictionaries you can specify: `redpanda.cluster` and `redpanda.node`. For more information, see [Cluster Configuration Properties](../../../../../../reference/cluster-properties) and [Node Configuration Properties](../../../../../../reference/node-properties). 

```bash
export JSONDATA='{"cluster":{"auto_create_topics_enabled":"true"},"node":{"developer_mode":"false"}}'
ansible-playbook ansible/<playbook-name>.yml --private-key artifacts/testkey -e redpanda="${JSONDATA}"
```

:::note

> Adding whitespace breaks configuration merging. 

Reviewer: it might be worth specifying that adding whitespace *to the JSON* will break things

```bash
ansible-playbook --private-key ~/.ssh/id_rsa ansible/<playbook-name>.yml -i hosts.ini -e redpanda_install_status=latest 
```

It's possible to upgrade clusters where SASL authentication has been turned on. For this, you need to additionally specify the `redpanda_rpk_opts` variable to include to username and password or a superuser or appropriately privileged user. For example:

```bash
ansible-playbook --private-key ~/.ssh/id_rsa ansible/<playbook-name>.yml -i hosts.ini --extra-vars=redpanda_install_status=latest --extra-vars "{
\"redpanda_rpk_opts\": \"--user ${MY_USER} --password ${MY_USER_PASSWORD}\"
}"
```

Similarly, you can put the `redpanda_rpk_opts` into a YAML file protected with Ansible vault.

```bash
ansible-playbook --private-key ~/.ssh/id_rsa ansible/<playbook-name>.yml -i hosts.ini --extra-vars=redpanda_install_status=latest --extra-vars @vault-file.yml --ask-vault-pass
```

##### Migrate to the AWS module

Redpanda previously delivered AWS code as simple in-place Terraform code, but it is now converted to a module. To migrate from the old AWS code:

1. Download the new configuration.
2. Run the following:
   
```bash

> terraform state list | while read -r line ; do terraform state mv "$line" "module.redpanda-cluster.$line"; done

Reviewer: This probably shouldn't be in the official docs as it is a hacky and jank solution to a temporary issue. While it will work most of the time, if someone has a different config or more stuff in their state file than just the stuff we created it could easily cause serious havoc
